# 01、02无关的一些小知识


冯诺伊曼体系的计算机有两大特征，一是将存储设备与中央处理器分开，二是数据以二进制方式编码


玛雅人是二十进制，按照玛雅人的历法，公元2012年是太阳纪的最后一年，2013年是新的太阳纪的开始，所以被以讹传讹为了2012年是玛雅人所说的世界末日。

# 03、04 变量的定义、运算符

**规则1：**变量名通常使用**小写英文字母**，多个单词用**下划线**进行连接。

**注意：**python中的字母指的是Unicode字符，**中文**、日文、希腊字母都可以作为变量名中的字符

**规则2：**受保护的变量用单个下划线开头。

**规则3：**私有的变量用两个下划线开头。

**规则4：**变量名一般用小写字母，中间用下划线连接。

### 求模与整除

In [ ]:
print(321 % 100) #求模，也就是求余数
#记忆法，%有两个圆圈，所以表示多余的，求余数
print(321 // 100) #整除，也就是只保留整数位，小数就不要了。

### 检查变量类型

In [2]:
a = 123
print(type(a))
# 要经常用type函数来对变量类型进行检查

<class 'int'>


### print函数的使用

In [8]:
f = float(input('请输入华氏温度:'))
c = (f-32)/1.8
print('%.1f华氏度 = %.1f摄氏度' % (f,c))

'''
%.1f 是一个占位符，后面用float变量来代替
%d   int占位符
%s   str字符串占位符
这种方式不利于记忆，我的感觉是不要用
'''

print(f"{f:.1f}华氏度 = {c:.1f}摄氏度") 
### 这个是非常关键的，即使是f类的print，也可以用.1f，之前我一直忽视。

'''
经典的f" 占位符模式，注意对于float变量，可以用:.1f来进行限定位数
下面还有例子
'''

请输入华氏温度:32
32.0华氏度 = 0.0摄氏度
32.0华氏度 = 0.0摄氏度


In [3]:
number = 3.1415926
print(f"数字为{number:.2f}")
print('%.3f就是我们要输出的内容。'% number) # 这种方式不好记，不要用

数字为3.14
3.142就是我们要输出的内容。


# 05 分支结构

In [5]:
#练习，请用户输入用户名和密码，如果是admin和123456则显示正确，否则显示错误
username = input('username:')
password = input('password:')

if username == 'admin' and password == '123456':
    print('correct.')
else:
    print('wrong')

username:admin
password:123456
correct.


代码嵌套的层次如果很多，会严重影响代码的可读性，所以更为扁平化的结果一般来说是更好的选择。

Zen of Python Flat is better than nested.

# 06 循环结构

如果明确知道**循环的次数**，推荐使用for-in循环

In [6]:
# 练习 用for-in循环实现1~100之间偶数的和
total = 0
for i in range(2,101,2): #注意(1,101,2) (2,101,2)都可以，因为range包含第一位数
    total  += i
print(total)

2500


如果**不知道要循环的次数**，就使用while循环

In [1]:
# 练习猜数字
# 游戏的规则是：计算机出一个1到100之间的随机数，玩家输入自己猜的数字，计算机给出对应的提示信息（大一点、小一点或猜对了），如果玩家猜中了数字，计算机提示用户一共猜了多少次，游戏结束，否则游戏继续。
import random
answer = random.randint(1,100)
count = 0

while True:
    number = int(input('输入一个1-100的数字：'))
    count += 1
    if number < answer:
        print('再大一点')
    elif number > answer:
        print('再小一点')
    else:
        print('猜对了，答案是%d.'% answer)
        print('一共猜了%d次。'% count)
        break # while遇到了break才会完全结束循环，注意与continue区分

输入一个1-100的数字：23
再小一点
输入一个1-100的数字：16
再小一点
输入一个1-100的数字：8
再大一点
输入一个1-100的数字：12
再大一点
输入一个1-100的数字：15
再小一点
输入一个1-100的数字：14
猜对了，答案是14.
一共猜了6次。


### *break与continue
**break**可以结束循环，但是只能结束它所在的循环，这一点在**嵌套循环结构**中有所体现。

如果不涉及嵌套循环，break就是很单纯的结束while循环，这样理解就可以了

**continue**用于放弃**本次循环的后续代码**，直接进入**下一次循环**。

In [4]:
#练习 循环嵌套的99乘法表
for i in range(1,10):
    for j in range(1,10):
        print(f'{i}*{j}={i*j}', end='\t') #attention!!
    print() #功能是实现换行输出

1*1=1	1*2=2	1*3=3	1*4=4	1*5=5	1*6=6	1*7=7	1*8=8	1*9=9	
2*1=2	2*2=4	2*3=6	2*4=8	2*5=10	2*6=12	2*7=14	2*8=16	2*9=18	
3*1=3	3*2=6	3*3=9	3*4=12	3*5=15	3*6=18	3*7=21	3*8=24	3*9=27	
4*1=4	4*2=8	4*3=12	4*4=16	4*5=20	4*6=24	4*7=28	4*8=32	4*9=36	
5*1=5	5*2=10	5*3=15	5*4=20	5*5=25	5*6=30	5*7=35	5*8=40	5*9=45	
6*1=6	6*2=12	6*3=18	6*4=24	6*5=30	6*6=36	6*7=42	6*8=48	6*9=54	
7*1=7	7*2=14	7*3=21	7*4=28	7*5=35	7*6=42	7*7=49	7*8=56	7*9=63	
8*1=8	8*2=16	8*3=24	8*4=32	8*5=40	8*6=48	8*7=56	8*8=64	8*9=72	
9*1=9	9*2=18	9*3=27	9*4=36	9*5=45	9*6=54	9*7=63	9*8=72	9*9=81	


In [6]:
# 练习 输入两个正整数，计算它们的最大公约数和最小公倍数 是for和if的组合拳练习

x = int(input('x ='))
y = int(input('y ='))
for factor in range(x,0,-1): # 这个属于range的妙用，从x到1，-1递减
    if x % factor == 0 and y % factor == 0:
        print(f"{x}和{y}最大公因数为：{factor}")
        print(f"{x}和{y}最小公倍数为：{x * y // factor}")
        break

x =45
y =84
45和84最大公因数为：3
45和84最小公倍数为：1260


# 07 分支和循环结构的应用

大部分初学者在学习了分支和循环结构后都能理解它们的用法，但是遇到实际问题时自己写不出来

不要沮丧，这是因为你的**练习量**还远远**没有达到可以让你随心所欲**的写出代码的程度。

练习！练习！练习！只有练习，才能真正熟练掌握代码技巧

In [7]:
# 练习 找出所有水仙花数 三位数，每个位上的立方之和等于这个数本身

for number in range(100,1000):
    gewei = number % 10
    shiwei = number // 10 % 10 #这一步非常精妙，还是值得好好品一品的
    baiwei = number // 100
    if number == gewei ** 3 + shiwei ** 3 + baiwei ** 3:
        print(number)

153
370
371
407


In [11]:
# 练习，将一个不知道多少位的正整数翻转，比如12345变成54321

number = int(input('num = '))
reversed_num = 0
while number > 0:
    reversed_num = reversed_num * 10 + number % 10
    number //= 10
print(reversed_num)
'''
这个设计非常的精妙，值得细细品味
number % 10 取末位数，并且把这个末位数保存下来
取完末位数之后之间用 //10 把最后一位给削去
太神奇了！一定要反复品味！！
'''

num = 1857845
5487581


In [13]:
# 练习 百钱百鸡问题
# 公鸡5元1个，母鸡3元1个，小鸡1元3个，用100元买100只鸡，问各有多少只？

# x为公鸡、y为母鸡、z为小鸡
for x in range(21):
    for y in range(33):
        z = 100 - x - y
        if z % 3 == 0 and 5 * x + 3 * y + z / 3 == 100:
            print(f'公鸡：{x}只，母鸡：{y}只，小鸡：{z}只')

公鸡：0只，母鸡：25只，小鸡：75只
公鸡：4只，母鸡：18只，小鸡：78只
公鸡：8只，母鸡：11只，小鸡：81只
公鸡：12只，母鸡：4只，小鸡：84只


In [16]:
# 练习 斐波那契数列
a, b = 0, 1
for _ in range(1,21):
    a, b = b, a+b #这种情况和拆开写的效果还真不一样
    print(a, end='\t')

1	1	2	3	5	8	13	21	34	55	89	144	233	377	610	987	1597	2584	4181	6765	

In [18]:
# 练习 打印100以内的素数 只能被1和自身整除的正整数（不包括1）

print('以下数字为2-99（包括2和99）间的素数：')
for num in range(2,100):
    is_prime = True
    for item in range(2,num):
        if num % item == 0:
            is_prime = False
            break
    if is_prime == True:
        print(num, end='\t')

以下数字为2-99（包括2和99）间的素数：
2	3	5	7	11	13	17	19	23	29	31	37	41	43	47	53	59	61	67	71	73	79	83	89	97	

### 循环和分支的毕业考试 CRAPS赌博游戏


**规则** ：玩家第一次摇骰子如果摇出了7点或11点，玩家胜；玩家第一次如果摇出2点、3点或12点，庄家胜；玩家如果摇出其他点数则玩家继续摇骰子，如果玩家摇出了7点，庄家胜；如果玩家摇出了第一次摇的点数，玩家胜；其他点数玩家继续摇骰子，直到分出胜负。


In [2]:
from random import randint
money = 1000 

while money > 0:
    print(f'你的总资产为：{money}元')
    while True: # 下注系统
        debt = int(input('请下注：'))
        if debt > money:
            print('下注金额超过了总资产，请重新下注。')
            continue
        elif 0 < debt <= money:
            break #跳出下注系统
    
    #第一次摇骰子
    go_on = False
    first = randint(1,6) + randint(1,6) # 模拟第一次摇骰子
    print(f'玩家首次投掷为{first}点')
    if first == 7 or first == 11:
        print('玩家胜')
        money += debt
    elif first == 2 or first == 3 or first == 12:
        print('庄家胜')
        money -= debt
    else:
        go_on = True
    #第一次做的时候有疑问：这个地方怎么才能搞出那种不运行后面代码的机制呢
    #看完之后可以用一个bool类型的simbol来搞，哈哈这一点真的学习体会到了
    
    while go_on == True:
        current = randint(1,6) + randint(1,6) # 模拟后面透骰子
        print(f'玩家投掷为{current}点')
        if current == 7:
            print('庄家胜')
            money -= debt
            break
        if current == first:
            print('玩家胜')
            money += debt
            break

print('你破产了，游戏结束')

你的总资产为：1000元
请下注：1000
玩家首次投掷为6点
玩家投掷为7点
庄家胜
你破产了，游戏结束


上面这个例子学到最重要的东西就是这个*go_on=True* 其实类似前面的*is_prime*

# 08 常用数据结构之列表

In [4]:
#练习 将一个骰子扔6000次，统计每个点数出现的次数
# 如果用下面这种方式来做，会感觉非常的笨重，这个时候就要引入列表啦。
from random import randint

time1=0
time2=0
time3=0
time4=0
time5=0
time6=0

for drop in range(6000):
    current = randint(1,6)
    if current == 1:
        time1 += 1
    elif current == 2:
        time2 += 1
    elif current == 3:
        time3 += 1
    elif current == 4:
        time4 += 1
    elif current == 5:
        time5 += 1
    elif current == 6:
        time6 += 1

print(f"1点出现了{time1}次。")
print(f"2点出现了{time2}次。")
print(f"3点出现了{time3}次。")
print(f"4点出现了{time4}次。")
print(f"5点出现了{time5}次。")
print(f"6点出现了{time6}次。")

1点出现了1004次。
2点出现了1004次。
3点出现了974次。
4点出现了1001次。
5点出现了993次。
6点出现了1024次。


### 列表的两种定义方式

In [5]:
# 方法一，直接手工定义
list1 = [1,2,3,4,5]
list2 = ['apple', 'pencil', 'ol']

# 方法二，用list函数将一些序列转变为列表
list1 = list(range(1,10))
print(list1)
# 注意，如果直接print(range(1,10))，输出的结果仍旧是range(1,10)，这个输出基本上来说毫无价值
list2 = list('hello')
print(list2)

[1, 2, 3, 4, 5, 6, 7, 8, 9]
['h', 'e', 'l', 'l', 'o']


### 列表的一些常用操作

In [11]:
items1 = [1,2,3,4,5,6]
items2 = [2,4,5,6,7,8,9]

# 列表的拼接
items3 = items1 +items2
print(items3)

# 列表的重复
items4 = items1 * 3 # 这里用[1,2,3,4,5,6] * 3效果是一样的
print(items4)

# 列表内的成员
print(100 in items3)
print(9 in items2) # 这种语法可以用来快速检索列表中有没有我们想要的元素

# 获取列表的元素个数
size = len(items2)
print(size)
print(len(items2)) #两种方法输出效果都是一样的。

[1, 2, 3, 4, 5, 6, 2, 4, 5, 6, 7, 8, 9]
[1, 2, 3, 4, 5, 6, 1, 2, 3, 4, 5, 6, 1, 2, 3, 4, 5, 6]
False
True
7
7


In [19]:
items1 = [1,2,3,4,5,6]
items2 = [2,4,5,6,7,8,9]

# 列表的索引
print(items1[0], items1[-2])
items1[1] = 100
print(items1[-5])

# 列表的切片，需要重点掌握，非常熟练才行啊。
print(items1[:5]) #意思是0，1，2，3，4位的数字，共5个
print(items1[-3:]) #意思是从倒数第三到倒数第一，共3个
print(items1[-2:-4:-1]) #这个是特别值得注意的
print(items1[::-2]) # 这个也很特殊，从后往前数，隔一位出一个数字

1 5
100
[1, 100, 3, 4, 5]
[4, 5, 6]
[5, 4]
[6, 4, 100]


In [22]:
#列表的比较运算
items1 = list(range(5))
items2 = [0,1,2,3,4]
items3 = [0,1,2,3,3]
print(items1 == items2)
print(items1 == items3) #列表相等的要求是每一位都要完全相同

items3 = [3,2,1]
print(items1 <= items3) # 这个地方我没有搞懂，目前来看也没什么意义，就先放着不要管了。

True
False
True


### 列表元素遍历的两种方法

In [26]:
items = ['a','b','c','d']

for item in items: # 这个是直接抓取元素
    print(item, end='\t')

print()

for index in range(len(items)): # 这个是不抓取元素，而是抓取index，感觉稍微麻烦一点
    print(items[index], end='\t')

a	b	c	d	
a	b	c	d	

In [28]:
# 练习 将一个骰子扔6000次，统计每个点数出现的次数
# 这个是跟这一节最开始一样的问题，但是咱们用列表来搞

import random
counters = [0] * 6 # 这样做就比[0,0,0,0,0,0]要方便很多，也不容易数错，这些细节很重要
for _ in range(6000):
    drop = random.randint(1,6)
    counters[drop - 1] += 1

for out in range(1,7):
    print(f"{out}点共出现{counters[out-1]}次")

# 不要小看上面这个小例子哦，有很多细节在里面，要熟练掌握。

1点共出现1000次
2点共出现1027次
3点共出现932次
4点共出现1057次
5点共出现940次
6点共出现1044次


### 列表的方法method
添加和删除元素

In [36]:
items = ['love','peace','study','hard']

# append在尾部添加元素
items.append('append') # 注意这个调用方式，冷不丁你还真不一定能想起来的，呵呵
print(items)

# 使用insert在列表指定位置添加元素
items.insert(1, 'insert') #注意这个位置还是以0，1，2这样的形式，并且插入的元素刚好处于新列表的这个位置
print(items)
items.insert(-1, '-2') #注意这个地方-1不是最后一位，因为-1的话其实和append就一样了
#这个地方有一点点难记，我们也不多管了。
print(items)

['love', 'peace', 'study', 'hard', 'append']
['love', 'insert', 'peace', 'study', 'hard', 'append']
['love', 'insert', 'peace', 'study', 'hard', '-2', 'append']


In [42]:
items1 = ['love','peace','study','hard']
#我们需要注意到的是，append，remove这些method都对原始的list发生了修改，所以如果想保留原始list的话，需要弄一个copy备份列表
items2=items1.copy()
# 这样的话无论是对原始列表还是新列表进行操作，都不会互相影响。


# 删除指定的元素 remove 直接说明你要删哪个
items2.remove('peace')
print(items2)
items2=items1.copy()

# 删除指定位置的元素
items2.pop(0)
items2.pop(len(items)-1) #这样相当于把最后一个元素给删去了
print(items2)
items2=items1.copy()

# 清空列表中的元素，这个应该是没什么用的
items2.clear()
print(items2)

['love', 'study', 'hard']
['peace', 'study']
[]


### 元素的位置和次数

In [47]:
items = ['love','peace','study','hard','love','love']

# 查找元素的索引位置
print(items.index('love'))
print(items.index('love',2)) # 意思就是查找索引为2及以后的love的位置
print(items.index('love',5)) # 意思就是查找索引为5及以后的love的位置，只找出现的第一个

# 查找元素出现的次数
print(items.count('love'))
print(items.count('study'))

0
4
5
3
1


元素的排序和翻转

In [48]:
items = ['love','peace','study','hard','love','love']
items1 = items.copy()

# 排序
items1.sort()
print(items)
items1 = items.copy()

# 反转
items1.reverse()
print(items1)

['love', 'peace', 'study', 'hard', 'love', 'love']
['love', 'love', 'hard', 'study', 'peace', 'love']


列表的生成式

**强烈建议用生成式语法来创建列表** 节约资源，更加快速，代码优雅

In [54]:
# 创建一个由1到9的数字构成的列表
items1 = [x for x in range(1,10)]
print(items1)

# 创建一个由'hello world'中除空格和元音字母外的字符构成的列表
items2 = [x for x in 'hello world' if x not in ' aeiou'] #这个语句很有难度的，要理解
print(items2)

# 创建一个由个两个字符串中字符的笛卡尔积构成的列表
items3 = [x + y for x in 'ABC' for y in '12'] # 卧槽，这个很骚
print(items3)

[1, 2, 3, 4, 5, 6, 7, 8, 9]
['h', 'l', 'l', 'w', 'r', 'l', 'd']
['A1', 'A2', 'B1', 'B2', 'C1', 'C2']


嵌套列表

**记住，现阶段不能用[[0] * 3] * 5]这种方式创建嵌套列表**

In [55]:
# 练习 保存5个学生3门课程的成绩

scores = [[0]*3 for _ in range(5)] # 这种方式非常的关键，创建嵌套列表用的，一定要熟练掌握。

scores[1][0] = 95 #一级结构是五个学生，二级结构是他们的分数
print(scores)

[[0, 0, 0], [95, 0, 0], [0, 0, 0], [0, 0, 0], [0, 0, 0]]


# 09 元组
元组类型的变量一旦定义，其中的元素不能再添加或删除，元素的值也不能修改

In [6]:
t1 = (30,20,15)
t2 = ('jiba', 'nice', 'finish')

print(len(t1), len(t2))

# 也可以通过索引找到元组中的元素
print(t1[-1], t2[1])

# 遍历元组中的元素
for item in t2:
    print(item)
    
# 判断元组中是否有所需要的元素
print(100 in t1)
print('jiba' in t2)

# 将两个元组拼接起来
t3 = t1 + t2

# 切片操作
print(t3[:2])
print(t3[::2])

3 3
15 nice
jiba
nice
finish
False
True
(30, 20)
(30, 15, 'nice')


注意，如果元组中只有一个元素，则需要在这个元素后加一个**逗号**

比如('hello',) ('hello')

前者是元组，而后者不是，只代表用括号括起来，实际上还是str的形式

In [11]:
#元组的作用，打包和解包
a = 1, 10, 100
print(type(a), a)

i, j, k = a #用这种解包的方式来获取元组中的元素还是非常有用的，不用一个个调取赋值
print(i, j, k)

# 注意，如果变量ijk的个数小于元组内的个数，会报错，有一个办法是用星号变量代表，变为列表
a1 = (1,2,3,4,5,6)
i,j,*k = a1
print(i,j,k)
i,*j = a1
print(i,j)
i,j,k,l,m,n,*o = a1 # 多重变量也可以是空列表
print(i,j,k,l,m,n,o)

<class 'tuple'> (1, 10, 100)
1 10 100
1 2 [3, 4, 5, 6]
1 [2, 3, 4, 5, 6]
1 2 3 4 5 6 []


这种解包语法非常常用，用来获取我们需要的元素

In [12]:
a, b, *c = range(1, 10)
print(a, b, c)
a, b, c = [1, 10, 100]
print(a, b, c)
a, *b, c = 'hello'
print(a, b, c)

1 2 [3, 4, 5, 6, 7, 8, 9]
1 10 100
h ['e', 'l', 'l'] o


元组的一个重要作用就是交换两个元素的值

In [13]:
a = 1
b = 'hello'
a,b = b,a
#这个地方的原理是 a,b = t = b,a
#首先a,b打包为一个元组t(1,'hello')，然后这个元组t解包给b和a两个变量，实现了调换
#非常非常方便好用!!!!!
print(a,b)

hello 1


元组与列表之间可以互相转换

In [14]:
info = ('a','b','c')
info_list = list(info)

list1 = ['a','b','c']
list1_tuple = tuple(list1)

# 10 字符串

In [22]:
s1 = 'fuck'
s2 = "fuck you"
s3 = '''
fuck
you'''
print(s3, 1, end='') #end=''表示输出后不换行，即将默认的结束符更换为空字符


fuck
you 1

### 转义字符

\n 换行

\t 制表符

在最前面加r，则表示里面不包含转义字符。

In [24]:
s1 = '\'hello,world\''
# 这个例子中，\'表示输出单引号，防止与字符串的单引号冲突
print(s1)

s2 = '\\hello,world\\'
# 这里面的\\的意思就是输出\这个符号，第一个\相当于一个转义符，第二个是我们要真正输出的
print(s2)

'hello,world'
\hello,world\


### 字符串的运算

In [25]:
#非常基础的拼接与重复
s1 = 'hello' + ' ' + 'world'
print(s1)    # hello world
s2 = '!' * 3
print(s2)    # !!!
s1 += s2     # s1 = s1 + s2
print(s1)    # hello world!!!
s1 *= 2      # s1 = s1 * 2
print(s1)    # hello world!!!hello world!!!

hello world
!!!
hello world!!!
hello world!!!hello world!!!


**比较运算**（感觉没什么大用）

对于两个字符串类型的变量，可以直接使用比较运算符比较两个字符串的相等性或大小。需要说明的是，因为字符串在计算机内存中也是以二进制形式存在的，那么字符串的大小比较比的是**每个字符对应的编码的大小**。例如A的编码是65， 而a的编码是97，所以'A' < 'a'的结果相当于就是65 < 97的结果，很显然是True；而'boy' < 'bad'，因为第一个字符都是'b'比不出大小，所以实际比较的是第二个字符的大小，显然'o' < 'a'的结果是False，所以'boy' < 'bad'的结果也是False。如果不清楚两个字符对应的编码到底是多少，可以使用ord函数来获得，例如ord('A')的值是65，而ord('昊')的值是26122。

**is运算符**

用来比较两个变量对应的字符串对象的内存地址

简单来说，就是两个变量是否对应内存中的同一个字符串

In [30]:
s1 = 'hello world'
s2 = 'hello world'
s3 = s2
print(s1 is s2, s2 is s3)

s1 = 'a'
s2 = 'a'
s3 = s2
print(s1 is s2, s2 is s3) #注意，如果这个字符串是一个单一字符，那内存地址是比较特殊的。

False True
True True


成员运算

In [1]:
s1 = 'hello, world'
print('lo' in s1)
s2 = 'goodbye'
print(s2 in s1) 

True
False


In [2]:
#获取字符串长度
s1 = 'hello, world'
print(len(s1))
print(len('goodbye my lover')) #注意，统计字符串长度时，空格也算在里面

12
16


### 字符串索引
注意，由于字符串是不可变类型，所以**不能**通过索引运算**修改字符串中的字符**

但是可以通过replace的方法来修改字符串中的字符

In [3]:
s = 'abc123456'
N = len(s)

print(s[0],s[-N]) #first str

print(s[N-1],s[-1]) #last str

#切片操作
print(s[2:5]) # output 2 3 4

print(s[-7:-4]) #同样也是正向切片

print(s[:4]) # output 0123

print(s[-N:-2]) #第一个到倒数第二个

print(s[2::2]) #把：看成一种分割符就好了，帮助理解

print(s[-7::2])

print(s[-1:-7:-2]) #反向切片 注意反向切片的最后一个步长一定要为负

print(s[::-1])

a a
6 6
c12
c12
abc1
abc1234
c246
c246
642
654321cba


In [28]:
# replace方法
s = 'hello world'
print(s.replace('o', '@'))
print(s.replace('o', '@', 1)) # 第三个参数是指定替换的次数
#其实这个方法还是有局限性的，但是可以用

hell@ w@rld
hell@ world


循环遍历每个字符，两种方式

In [6]:
s1 = 'hello'
for index in range(len(s1)):
    print(s1[index], end=' ')

h e l l o 

In [8]:
s1 = 'hello'
for ch in s1:
    print(ch, end=' ')

h e l l o 

### 字符串的一些方法

In [12]:
s1 = 'hello, world, i am your father'
print(s1.capitalize()) # 整个字符串的唯一首字母大写
print(s1.title()) # 每一个单词的首字母都大写
print(s1.upper()) # 所有字母全部放大

s2 = 'GooDBYe'
print(s2.lower())

Hello, world, i am your father
Hello, World, I Am Your Father
HELLO, WORLD, I AM YOUR FATHER
goodbye


In [14]:
# 在一个字符串中找到另一个字符串的索引
# 这两个的功能完全一致
s1 = 'hello, world!'
print(s1.find('or'))
# 找到了返回字符串中另一个字符串首字符的索引
print(s1.index('or'))

8
8


In [ ]:
#startwith和endwith来判断字符串的开头结尾
#感觉其实都没什么用
s1 = 'hello, world!'

# startwith方法检查字符串是否以指定的字符串开头返回布尔值
print(s1.startswith('He'))    # False
print(s1.startswith('hel'))   # True
# endswith方法检查字符串是否以指定的字符串结尾返回布尔值
print(s1.endswith('!'))       # True

s2 = 'abc123456'

# isdigit方法检查字符串是否由数字构成返回布尔值
print(s2.isdigit())    # False
# isalpha方法检查字符串是否以字母构成返回布尔值
print(s2.isalpha())    # False
# isalnum方法检查字符串是否以数字和字母构成返回布尔值
print(s2.isalnum())    # True

### 字符串的位置

In [24]:
s = ' hello, world '
# center方法以宽度20将字符串居中并在两侧填充*
print(s.center(30, '*'))
# rjust方法以宽度20将字符串右对齐并在左侧填充空格
print(s.rjust(20))
# ljust方法以宽度20将字符串左对齐并在右侧填充~
print(s.ljust(20,'-'))

# 在字符串的左侧补零（这个非常有用！！！！）
print(s.zfill(30))
print('-32'.zfill(5))

******** hello, world ********
       hello, world 
 hello, world ------
0000000000000000 hello, world 
-0032


| 变量值      | 占位符     | 格式化结果    | 说明 |
| ----------- | ---------- | ------------- | ---- |
| `3.1415926` | `{:.2f}`   | `'3.14'`      | 保留小数点后两位 |
| `3.1415926` | `{:+.2f}`  | `'+3.14'`       | 带符号保留小数点后两位 |
| `-1`        | `{:+.2f}`  | `'-1.00'` | 带符号保留小数点后两位 |
| `3.1415926` | `{:.0f}`   | `'3'` | 不带小数 |
| `123`       | `{:0>10d}` | `'0000000123'` | 左边补`0`，补够10位 |
| `123`       | `{:x<10d}` | `'123xxxxxxx'` | 右边补`x` ，补够10位 |
| `123`       | `{:>10d}`  | `'       123'` | 左边补空格，补够10位 |
| `123`       | `{:<10d}` | `'123       '` | 右边补空格，补够10位 |
| `123456789` | `{:,}`     | `'123,456,789'` | 逗号分隔格式 |
| `0.123`     | `{:.2%}`   | `'12.30%'`    | 百分比格式 |
| `123456789` | `{:.2e}`   | `'1.23e+08'`  | 科学计数法格式 |

### 字符串的修剪

In [27]:
s = '   jackfrued@126.com  \t\r\n'
# strip方法获得字符串修剪左右两侧空格之后的字符串 注：\r表示回车
print(s.strip())    
print(s.lstrip()) #去掉前面的空格
print(s.rstrip()) #去掉后面的空格

jackfrued@126.com
jackfrued@126.com  	

   jackfrued@126.com


### 字符串的拆分、合并

In [30]:
s = 'I love you' # 以空格为标准进行字符串的拆分
words = s.split()
print(words)
print('+'.join(words)) # 这个join的语法也是非常神奇，但是应该非常好用

['I', 'love', 'you']
I+love+you


# 11 数据结构之集合

1. 无序性：一个集合中，每个元素的地位都是相同的，元素之间是无序的。
2. 互异性：一个集合中，**任何两个元素都是不相同的**，即元素在集合中只能出现一次。
3. 确定性：给定一个集合和一个任意元素，该元素要么属这个集合，要么不属于这个集合，二者必居其一，不允许有模棱两可的情况出现


很显然，python中的集合**不支持索引运算**

**集合的成员运算在性能上要优于列表的成员**，仅记住这个结论就可以，跟内存相关，不做讨论

创建方式为{}，**至少有一个元素**，如果没有元素则为**空字典**

In [38]:
# 创建集合的方式
set1 = {1,2,3,1,2,3}
print(set1) #重复的元素会自动筛除
print(len(set1))

set2 = set('hello') #重复的元素l会删除
print(set2)

#把列表转换为集合
list1 = [1,2,2,2,1,3,4,5]
set3 = set(list1)
print(set3)

#创建集合的生成式语法，类似列表的生成式语法
set4 = {num for num in range(1,20) if num %3 == 0 or num % 5 == 0}
print(set4)

#遍历集合中的元素
for elem in set4:
    print('fuck', elem, end=' ')

{1, 2, 3}
3
{'o', 'h', 'e', 'l'}
{1, 2, 3, 4, 5}
{3, 5, 6, 9, 10, 12, 15, 18}
fuck 3 fuck 5 fuck 6 fuck 9 fuck 10 fuck 12 fuck 15 fuck 18 

需要注意的是，集合中的元素必须是hashable类型。

通常不可变类型都是hashable类型，如整数、浮点、字符串、元组等，而可变类型都不是hashable类型，因为可变类型无法确定唯一的ID值，所以也就不能放到集合中。集合本身也是可变类型，所以集合不能够作为集合中的元素，这一点在使用集合的时候一定要注意。

### 集合的运算

In [41]:
#in 和 not in
set1 = {1,2,3,4,3,2,}
print(10 in set1)
print(1 in set1)
print(5 not in set1)

False
True
True


In [51]:
#交集与并集
set1 = {1, 2, 3, 4, 5, 6, 7}
set2 = {2, 4, 6, 8, 10}

#交集
#方法一：&运算符
print('交集', set1 & set2)
#方法二：intersection方法
print(set1.intersection(set2))

#并集
# 方法一：使用|运算符
print('并集', set1 | set2)
# 方法二：使用union方法
print(set1.union(set2))

#差集
# 方法一：使用 - 运算符
print('差集', set1 - set2)
print(set2-set1) #注意，这两个是有区别的
# 方法二：使用difference方法
print(set1.difference(set2))

#对称差 就相当于（并集-交集）
# 方法一：使用^运算符
print('对称差', set1 ^ set2)
# 方法二：使用symmetric_difference方法
print(set1.symmetric_difference(set2))
# 方法三：相当于并集减去交集
print((set1 | set2) - (set1 & set2))

交集 {2, 4, 6}
{2, 4, 6}
并集 {1, 2, 3, 4, 5, 6, 7, 8, 10}
{1, 2, 3, 4, 5, 6, 7, 8, 10}
差集 {1, 3, 5, 7}
{8, 10}
{1, 3, 5, 7}
对称差 {1, 3, 5, 7, 8, 10}
{1, 3, 5, 7, 8, 10}
{1, 3, 5, 7, 8, 10}


In [53]:
#集合运算的复合赋值运算
set1 = {1,3,5,7}
set2 = {2,4,6}

set1 |= set2
print(set1)

set3 = {3,6,9}
set1 &= set3
print(set1)

{1, 2, 3, 4, 5, 6, 7}
{3, 6}


In [56]:
#<用来计算真子集，<=用来计算子集
set1 = {1,3,5}
set2 = {1,2,3,4,5}
print(set1 < set2, set1 <=set2)

#通过issubset方法来进行子集判断
print(set1.issubset(set2))

#通过issuperset进行超集判断
print(set1.issuperset(set2))

True True
True
False


### 集合添加或删除元素

In [65]:
set1 = set()

#通过add来进行赋值
set1.add(33)
set1.add(44)
print(set1)
set1.update({1,10,2,33,44}) #等同于set1 |= set2
print(set1)

#通过discard方法来删除指定元素
set1.discard(10)
set1.discard(10) #这种方法重复使用不会报错，也就是说这个元素已经没有了，

#通过remove方法来删除指定元素
if 33 in set1:
    set1.remove(33)
print(set1)

#通过pop的方法从集合中随机删除一个元素并返回该元素
print(set1.pop())

{33, 44}
{33, 1, 2, 10, 44}
{1, 2, 44}
1


In [66]:
#判断两个集合有没有相同的元素
set1 = {'Java', 'Python', 'Go', 'Kotlin'}
set2 = {'Kotlin', 'Swift', 'Java', 'Objective-C', 'Dart'}
set3 = {'HTML', 'CSS', 'JavaScript'}
print(set1.isdisjoint(set2))    # False
print(set1.isdisjoint(set3))

False
True


frozenset就是跟set集合基本是一样的，但是frozenset类似tuple,里面的元素不能修改

In [67]:
set1 = frozenset({1,3,5,7})
set2 = frozenset(range(1,6))
print(set1 - set2)

frozenset({7})


# 12字典

In [2]:
dict1 = {
    'name':'luogabo',
    'age':12,
    'weight':80,
    'wife':'林志玲',
    'tel':15651758852
}
print(dict1)

{'name': 'luogabo', 'age': 12, 'weight': 80, 'wife': '林志玲'}


In [7]:
# 用dict函数（构造器）来定义字典
dict1 = dict(name='罗尬博', age=55, weight=80, home='同人路8路')
print(dict1)

# 用python内置zip函数压缩两个序列并创建字典，感觉没什么用
items1 = dict(zip('abcde','12345'))
print(items1)
items2 = dict(zip('abcde', range(1,10)))
print(items2)

{'name': '罗尬博', 'age': 55, 'weight': 80, 'home': '同人路8路'}
{'a': '1', 'b': '2', 'c': '3', 'd': '4', 'e': '5'}
{'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5}


### 用字典生成式语法创建字典

In [13]:
items3 = {x: x ** 3 for x in range(1,6)}
print(items3)

{1: 1, 2: 8, 3: 27, 4: 64, 5: 125}


In [16]:
# 用len函数来查询字典有多少个键值对
person = {'name':'fuck', 'age':'you', 'weight':32}
print(len(person))

for key in person:
    print(key) #这样只能输出key值，输出键值对的方式也特别简单，看下面

3
name
age
weight


字典的**索引**不是序号，而是键值对中的键

字典中的键必须是不可变类型，例如整数、浮点数、字符串、元组，显然列表list和集合set是不能作为字典中的键的。

In [18]:
person = {'name': '王大锤', 'age': 55, 'weight': 60, 'office': '科华北路62号'}

#修改键的值
person['age'] = 25

#增加键值对
person['tel'] = '13122334455'
print('tel' in person)
print(len(person))

#遍历字典所有的键值对
for key in person:
    print(f'{key}:{person[key]}')

True
5
name:王大锤
age:25
weight:60
office:科华北路62号
tel:13122334455


In [4]:
#嵌套字典，也就是字典的值也是一个字典，而不是一个单独的变量
students = {
    220193084:{'name':'liuyuhao', 'sex':'man', 'age':25, 'place':'辽宁大连'},# 注意这里需要有逗号，很容易忽略
    220193012:{'name':'123', 'sex':'man', 'age':24, 'place':'湖南长沙'},
    220193024:{'name':'456', 'sex':'man', 'age':27, 'place':'四川成都'},
}

# 通过get方法通过键获取对应的值
print(students.get(220193012))
print(students.get(220191111, {'name':'未注册'}))
#这个非常关键，如果get不到相应的键，就会输出这个默认值

# 也可以通过这种方式来获取对应的值
print(students[220193012]) 
print(students[220193012]['name'])

#获取字典中的所有键
print(students.keys())

#获取字典中的所有值
print(students.values())

#获取字典中的所有键值对
print(students.items())

#对字典中的所有键值对进行循环遍历
for key, value in students.items():
    print(key, '--->', value)

{'name': '123', 'sex': 'man', 'age': 24, 'place': '湖南长沙'}
{'name': '123', 'sex': 'man', 'age': 24, 'place': '湖南长沙'}
123
dict_keys([220193084, 220193012, 220193024])
dict_values([{'name': 'liuyuhao', 'sex': 'man', 'age': 25, 'place': '辽宁大连'}, {'name': '123', 'sex': 'man', 'age': 24, 'place': '湖南长沙'}, {'name': '456', 'sex': 'man', 'age': 27, 'place': '四川成都'}])
dict_items([(220193084, {'name': 'liuyuhao', 'sex': 'man', 'age': 25, 'place': '辽宁大连'}), (220193012, {'name': '123', 'sex': 'man', 'age': 24, 'place': '湖南长沙'}), (220193024, {'name': '456', 'sex': 'man', 'age': 27, 'place': '四川成都'})])
220193084 ---> {'name': 'liuyuhao', 'sex': 'man', 'age': 25, 'place': '辽宁大连'}
220193012 ---> {'name': '123', 'sex': 'man', 'age': 24, 'place': '湖南长沙'}
220193024 ---> {'name': '456', 'sex': 'man', 'age': 27, 'place': '四川成都'}


In [11]:
students = {
    220193084:{'name':'liuyuhao', 'sex':'man', 'age':25, 'place':'辽宁大连'},# 注意这里需要有逗号，很容易忽略
    220193012:{'name':'123', 'sex':'man', 'age':24, 'place':'湖南长沙'},
    220193024:{'name':'456', 'sex':'man', 'age':27, 'place':'四川成都'},
}

#使用pop方法通过键删除对应的键值对
stu1 = students.pop(220193084)
print(stu1)

#使用popitem方法删除字典中最后一组键值对并返回对应的二元组
key, value = students.popitem()
print(key, value)

#setdefault 如果这个键在字典中存在，setfault返回原来这个键对应的值
#否则就在最后添加这个键值对
result = students.setdefault(220193084,{'name':'new'})
result2 = students.setdefault(220193012,123) #因为220193012是存在的，所以后面的赋值123相当于无效
print(students)
print(result2)


#使用update更新字典元素，相同的键会用新值覆盖掉旧值，不同的键会添加到字典中
#注意哦，可以用字典来更新字典，一下子量会非常大！
others = {
    220193084:{'name':'123','sex':'female'},
    220193012:{'name':'abc','sex':'male'},
}
students.update(others) #相当于用一个字典来更新另一个字典了
print(students)

{'name': 'liuyuhao', 'sex': 'man', 'age': 25, 'place': '辽宁大连'}
220193024 {'name': '456', 'sex': 'man', 'age': 27, 'place': '四川成都'}
{220193012: {'name': '123', 'sex': 'man', 'age': 24, 'place': '湖南长沙'}, 220193084: {'name': 'new'}}
{'name': '123', 'sex': 'man', 'age': 24, 'place': '湖南长沙'}
{220193012: {'name': 'abc', 'sex': 'male'}, 220193084: {'name': '123', 'sex': 'female'}}


字典的应用

In [1]:
#练习1 输入一段话，统计每个英文字母出现的次数
sentence = input('please input a sentence:')
counter = {}
for ch in sentence:
    if 'A' <= ch <= 'Z' or 'a' <= ch <= 'z':
        counter[ch] = counter.get(ch, 0) + 1 
        # 这个get(ch, 0)还是非常精妙的，如果第一次计数，就设置默认值为0，不错的

for key, value in counter.items():
    print(f'字母{key}出现了{value}次')

please input a sentence:cawer4fadf216e346qwd1asf1aeqwd6aSDf16ased1fq33wff
字母c出现了1次
字母a出现了6次
字母w出现了4次
字母e出现了4次
字母r出现了1次
字母f出现了7次
字母d出现了4次
字母q出现了3次
字母s出现了2次
字母S出现了1次
字母D出现了1次


In [2]:
#练习2 在一个字典中保存股票的代码和价格，找出股价大于100元的股票并创建一个新的字典
stocks = {
    'AAPL': 191.88,
    'GOOG': 1186.96,
    'IBM': 149.24,
    'ORCL': 48.44,
    'ACN': 166.89,
    'FB': 208.09,
    'SYMC': 21.29
}
stocks2 = {key: value for key, value in stocks.items() if value > 100}
# 哇这个语法用的相当娴熟了！牛逼！
print(stocks2)

{'AAPL': 191.88, 'GOOG': 1186.96, 'IBM': 149.24, 'ACN': 166.89, 'FB': 208.09}


简单总结

字典以**键值对的形式保存数据**，再通过键**索引**对应的值，非常有利于**数据检索**